cmd/object-lock.go

```diff
@@ -84,65 +186,51 @@ func enforceRetentionBypassForDelete(ctx context.Context, r *http.Request, bucke
 // For objects in "Governance" mode, overwrite is allowed if a) object retention date is past OR
 // governance bypass headers are set and user has governance bypass permissions.
 // Objects in compliance mode can be overwritten only if retention date is being extended. No mode change is permitted.
-func enforceRetentionBypassForPut(ctx context.Context, r *http.Request, bucket, object string, getObjectInfoFn GetObjectInfoFn, govBypassPerm APIErrorCode, objRetention *objectlock.ObjectRetention) (oi ObjectInfo, s3Err APIErrorCode) {
-	if globalWORMEnabled {
-		return oi, ErrObjectLocked
-	}
-
-	var err error
-	var opts ObjectOptions
-	opts, err = getOpts(ctx, r, bucket, object)
+func enforceRetentionBypassForPut(ctx context.Context, r *http.Request, bucket, object string, getObjectInfoFn GetObjectInfoFn, objRetention *objectlock.ObjectRetention) (ObjectInfo, APIErrorCode) {
+	byPassSet := objectlock.IsObjectLockGovernanceBypassSet(r.Header)
+	opts, err := getOpts(ctx, r, bucket, object)
 	if err != nil {
-		return oi, toAPIErrorCode(ctx, err)
+		return ObjectInfo{}, toAPIErrorCode(ctx, err)
 	}
-	oi, err = getObjectInfoFn(ctx, bucket, object, opts)
+
+	oi, err := getObjectInfoFn(ctx, bucket, object, opts)
 	if err != nil {
-		// ignore case where object no longer exists
-		if toAPIError(ctx, err).Code == "NoSuchKey" {
-			oi.UserDefined = map[string]string{}
-			return oi, ErrNone
-		}
 		return oi, toAPIErrorCode(ctx, err)
 	}
 
 	ret := objectlock.GetObjectRetentionMeta(oi.UserDefined)
-	// no retention metadata on object
-	if ret.Mode == objectlock.Invalid {
-		if _, isWORMBucket := globalBucketObjectLockConfig.Get(bucket); !isWORMBucket {
-			return oi, ErrInvalidBucketObjectLockConfiguration
-		}
-		return oi, ErrNone
-	}
-	t, err := objectlock.UTCNowNTP()
-	if err != nil {
-		logger.LogIf(ctx, err)
-		return oi, ErrObjectLocked
-	}
+	if ret.Mode.Valid() {
+		switch ret.Mode {
+		case objectlock.RetGovernance:
+			// Governance mode retention period cannot be shortened, if x-amz-bypass-governance is not set.
+			if !byPassSet {
+				if objRetention.RetainUntilDate.Before((ret.RetainUntilDate.Time)) {
+					return oi, ErrObjectLocked
+				}
+				// Retention period in future is allowed.
+				return oi, ErrNone
+			}
 
-	if ret.Mode == objectlock.Compliance {
-		// Compliance retention mode cannot be changed and retention period cannot be shortened as per
-		// https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lock-overview.html#object-lock-retention-modes
-		if objRetention.Mode != objectlock.Compliance || objRetention.RetainUntilDate.Before(ret.RetainUntilDate.Time) {
-			return oi, ErrObjectLocked
-		}
-		if objRetention.RetainUntilDate.Before(t) {
-			return oi, ErrInvalidRetentionDate
-		}
-		return oi, ErrNone
-	}
+			// Pass in relative days to additionally verify "object-lock-remaining-retention-days" policy if any.
+			days := int(math.Abs(objRetention.RetainUntilDate.Sub(ret.RetainUntilDate.Time).Hours()) / 24)
```

Comment:
@harshavardhana , seems to me the objRetention.RetainUntilDate should be relative to current time rather than previous retention setting (if any). 

https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lock-managing.html

Reply:
@poornas I am not able to understand where do you see that.

Reply:
>  Enables enforcement of an object relative to the remaining retention days

The remaining is relative to what was currently set @poornas

Reply:
ok, got it - was misled by the verbiage in the link posted above. Tested with the previous policy and it works as expected with the NumericGreaterThan condition. However the NumericLessThan condition allowed me to extend retention date well beyond the condition constraint.
In the same policy I posted before, changed the constraint to
```
    "NumericLessThan": {
          "s3:object-lock-remaining-retention-days": "10"
        }
```

and 
```
➜ mc cp /etc/issue myminio/lhold/abc
/etc/issue:        26 B / 26 B ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 3.63 KiB/s 0s➜  export aws s3api put-object-retention --bucket lhold --key abc --retention Mode=GOVERNANCE,RetainUntilDate="2020-06-13T18:20:00Z" --bypass-governance-retention  --endpoint-url http://localhost:9000 --profile kp1
➜  aws s3api put-object-retention --bucket lhold --key abc --retention Mode=GOVERNANCE,RetainUntilDate="2020-07-13T18:20:00Z" --bypass-governance-retention  --endpoint-url http://localhost:9000 --profile kp1
```
the second override of retention date to 2020-07-13 should have been disallowed because it exceeds the 10 day constraint set for remaining-days

Reply:
I don't think you are testing it right @poornas did you update the policy by setting the policy again because I can't reproduce it? 

```
~ mc stat myminio/lhold/hosts
Name      : hosts
Date      : 2020-04-04 10:44:19 PDT 
Size      : 282 B  
ETag      : af639b05c71d860f4638b509be41019c-1 
Type      : file 
Metadata  :
  Content-Type                       : application/octet-stream 
  X-Amz-Object-Lock-Mode             : GOVERNANCE 
  X-Amz-Object-Lock-Retain-Until-Date: 2020-06-06T18:20:00Z
```

```
~ {
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "Retentionperms",
      "Effect": "Allow",
      "Action": [
        "s3:BypassGovernanceRetention",
        "s3:PutObjectRetention"
      ],
      "Resource": [
        "arn:aws:s3:::*"
      ],
      "Condition": {
        "DateGreaterThan": {
          "s3:object-lock-retain-until-date": [
            "2020-05-31T10:00:00Z"
          ]
        },
        "NumericLessThan": {
          "s3:object-lock-remaining-retention-days": [
            10
          ]
        },
        "StringEquals": {
          "s3:object-lock-mode": [
            "GOVERNANCE"
          ]
        }
      }
    },
    {
      "Sid": "AllowAllBucketActionsRead",
      "Effect": "Allow",
      "Action": [
        "s3:GetBucketLocation",
        "s3:ListAllMyBuckets"
      ],
      "Resource": [
        "arn:aws:s3:::*"
      ]
    }
  ]
}
```

```
~ aws s3api put-object-retention --bucket lhold --key hosts --retention Mode=GOVERNANCE,RetainUntilDate="2020-07-06T18:20:00Z" --endpoint-url http://localhost:9000 --profile harsha --bypass-governance-retention

An error occurred (AccessDenied) when calling the PutObjectRetention operation: Access Denied.
```

Whereas less than 10days succeeded

```
~ aws s3api put-object-retention --bucket lhold --key hosts --retention Mode=GOVERNANCE,RetainUntilDate="2020-06-09T18:20:00Z" --endpoint-url http://localhost:9000 --profile harsha --bypass-governance-retention
```

``` 
~ mc stat myminio/lhold/hosts
Name      : hosts
Date      : 2020-04-04 10:44:19 PDT 
Size      : 282 B  
ETag      : af639b05c71d860f4638b509be41019c-1 
Type      : file 
Metadata  :
  Content-Type                       : application/octet-stream 
  X-Amz-Object-Lock-Retain-Until-Date: 2020-06-09T18:20:00Z 
  X-Amz-Object-Lock-Mode             : GOVERNANCE 
```